includes/breadcrumbs.inc

```diff
@@ -0,0 +1,236 @@
+<?php
+
+/**
+ * @file
+ * This file contains functions to create breadcrumbs using Solr.
+ */
+
+/**
+ * Implements callback_islandora_breadcrumbs_backends.
+ */
+function islandora_solr_islandora_breadcrumbs_backends_callable($object) {
+  $breadcrumbs = islandora_solr_get_breadcrumbs_recursive($object->id);
+  array_pop($breadcrumbs);
+  return $breadcrumbs;
+}
+
+/**
+ * Loads a map of collection hierarchy.
+ *
+ * @return array
+ *   The array has current pid as key pointing to an associative array of
+ *   'parent' => pid and 'label' => parent_label.
+ */
+function islandora_solr_load_crumbs() {
+  if (user_is_anonymous()) {
+    $cache_roles = 'anonymous';
+  }
+  else {
+    global $user;
+    $roles = $user->roles;
+    // Sort so we keep ordering consistent to reuse cache.
+    asort($roles);
+    $cache_roles = implode('_', $roles);
+  }
+
+  $crumb_cache = &drupal_static(__FUNCTION__);
+
+  $breadcrumbs = array();
+
+  if (!isset($crumb_cache)) {
+    $cache = cache_get(ISLANDORA_SOLR_BREADCRUMB_CACHE_NAME, 'cache_menu');
+    if (!$cache) {
+      // Nothing in cache so initialize it with these roles.
+      $fetch = $cache_roles;
+    }
+    else {
+      $crumb_cache = $cache->data;
+    }
+  }
+
+  if (!isset($crumb_cache[$cache_roles])) {
+    // No cache for this set of roles.
+    $fetch = $cache_roles;
+  }
+  else {
+    $breadcrumbs = $crumb_cache[$cache_roles];
+  }
+
+  if (!$breadcrumbs) {
+    $fetch = $cache_roles;
+  }
+
+  if (isset($fetch)) {
+    module_load_include('inc', 'islandora_solr_search', 'includes/query_processor');
+    $solr_build = new IslandoraSolrQueryProcessor();
+
+    $content_model = variable_get('islandora_solr_content_model_field', 'RELS_EXT_hasModel_uri_ms');
+    $collection_field = variable_get('islandora_solr_member_of_collection_field', 'RELS_EXT_isMemberOfCollection_uri_ms');
```

Comment:
@willtp87 That's a fair point. As anyone can define their structure as they see fit this requires more configuration.

The best solution is for someone to add a configuration variable that asks for all the possible parent Solr fields and then does a query for them all and figures out which one has a value in it. It can use the same recursion. Just without the cached map.

I'm going to close this PR and move this code to a small add-on module that I (and anyone with a simple structure) can use. Then I'll think about how to proceed for a general approach.